## A toolkit of Python Dictionaries

* The following Python Dictionaries will be used to aid our search

In [1]:
codon = {  # Maps an RNA triplet of nucelotides to a 1-letter Amino Acid Abbrevation 
    "AAA": 'K', "AAG": 'K', "AAC": 'N', "AAU": 'N',
    "AGA": 'R', "AGG": 'R', "AGC": 'S', "AGU": 'S',
    "ACA": 'T', "ACG": 'T', "ACC": 'T', "ACU": 'T',
    "AUA": 'I', "AUG": 'M', "AUC": 'I', "AUU": 'I',
    "GAA": 'E', "GAG": 'E', "GAC": 'D', "GAU": 'D',
    "GGA": 'G', "GGG": 'G', "GGC": 'G', "GGU": 'G',
    "GCA": 'A', "GCG": 'A', "GCC": 'A', "GCU": 'A',
    "GUA": 'V', "GUG": 'V', "GUC": 'V', "GUU": 'V',
    "CAA": 'Q', "CAG": 'Q', "CAC": 'H', "CAU": 'H',
    "CGA": 'R', "CGG": 'R', "CGC": 'R', "CGU": 'R',
    "CCA": 'P', "CCG": 'P', "CCC": 'P', "CCU": 'P',
    "CUA": 'L', "CUG": 'L', "CUC": 'L', "CUU": 'L',
    "UAA": '*', "UAG": '*', "UAC": 'Y', "UAU": 'Y',
    "UGA": '*', "UGG": 'W', "UGC": 'C', "UGU": 'C',
    "UCA": 'S', "UCG": 'S', "UCC": 'S', "UCU": 'S',
    "UUA": 'L', "UUG": 'L', "UUC": 'F', "UUU": 'F'
}

AminoAcid = { # Maps 1-letter Amino Acid Abbrevations to their full name
    'A': 'Alanine', 'C': 'Cysteine', 'D': 'Aspartic acid', 'E': 'Glutamic acid', 'F': 'Phenylalanine',
    'G': 'Glycine', 'H': 'Histidine', 'I': 'Isoleucine', 'K': 'Lysine', 'L': 'Leucine', 'M': 'Methionine',
    'N': 'Asparagine', 'P': 'Proline', 'Q': 'Glutamine', 'R': 'Arginine', 'S': 'Serine',
    'T': 'Theronine', 'V': 'Valine', 'W': 'Tryptophan', 'Y': 'Tyrosine', '*': 'STOP'
}

AminoAbbrv = { # Maps 1-letter Amino Acid Abbrevations to 3-letter Abbrevations
    'A': 'Ala', 'C': 'Cys', 'D': 'Asp', 'E': 'Glu', 'F': 'Phe', 'G': 'Gly', 'H': 'His', 'I': 'Ile',
    'K': 'Lys', 'L': 'Leu', 'M': 'Met', 'N': 'Asn', 'P': 'Pro', 'Q': 'Gln', 'R': 'Arg', 'S': 'Ser',
    'T': 'Thr', 'V': 'Val', 'W': 'Trp', 'Y': 'Tyr', '*': 'STP'    
}

## Now our Peptide sequence
We'll use the 1-letter Amino Acid abbreviations to represent our sequence, and create a dictionary that provides for every peptide in our chain the set of codons that could encode it.

In [2]:
TrimerCodes = {}
for key, code in codon.items():
    TrimerCodes[code] = TrimerCodes.get(code,[]) + [key]
for key in sorted(TrimerCodes.keys()):
    print("%1s (%13s): %s" % (key, AminoAcid[key], TrimerCodes[key]))

* (         STOP): ['UAA', 'UAG', 'UGA']
A (      Alanine): ['GCA', 'GCG', 'GCC', 'GCU']
C (     Cysteine): ['UGC', 'UGU']
D (Aspartic acid): ['GAC', 'GAU']
E (Glutamic acid): ['GAA', 'GAG']
F (Phenylalanine): ['UUC', 'UUU']
G (      Glycine): ['GGA', 'GGG', 'GGC', 'GGU']
H (    Histidine): ['CAC', 'CAU']
I (   Isoleucine): ['AUA', 'AUC', 'AUU']
K (       Lysine): ['AAA', 'AAG']
L (      Leucine): ['CUA', 'CUG', 'CUC', 'CUU', 'UUA', 'UUG']
M (   Methionine): ['AUG']
N (   Asparagine): ['AAC', 'AAU']
P (      Proline): ['CCA', 'CCG', 'CCC', 'CCU']
Q (    Glutamine): ['CAA', 'CAG']
R (     Arginine): ['AGA', 'AGG', 'CGA', 'CGG', 'CGC', 'CGU']
S (       Serine): ['AGC', 'AGU', 'UCA', 'UCG', 'UCC', 'UCU']
T (    Theronine): ['ACA', 'ACG', 'ACC', 'ACU']
V (       Valine): ['GUA', 'GUG', 'GUC', 'GUU']
W (   Tryptophan): ['UGG']
Y (     Tyrosine): ['UAC', 'UAU']


## How many ways to code Tyrocidine B1?

* Since most Amino Acids have multiple codon encodings, most peptide sequences can be codes in multiple ways.
* Let's figure out exactly how many ways there are to encode our little peptide.

In [3]:
TyrocidineB1 = "VKLFPWFNQY"
codes = 1
for residue in TyrocidineB1:
    print(residue, AminoAbbrv[residue], TrimerCodes[residue])
    codes *= len(TrimerCodes[residue])
print("%d possible sequences" % codes)

V Val ['GUA', 'GUG', 'GUC', 'GUU']
K Lys ['AAA', 'AAG']
L Leu ['CUA', 'CUG', 'CUC', 'CUU', 'UUA', 'UUG']
F Phe ['UUC', 'UUU']
P Pro ['CCA', 'CCG', 'CCC', 'CCU']
W Trp ['UGG']
F Phe ['UUC', 'UUU']
N Asn ['AAC', 'AAU']
Q Gln ['CAA', 'CAG']
Y Tyr ['UAC', 'UAU']
6144 possible sequences


## Some Helper functions

In [4]:
def revComp(dnaSeq):
    return ''.join([{'A':'T','C':'G','G':'C','T':'A'}[base] for base in reversed(dnaSeq)])

def proteinTranslation(dnaSeq):
    rnaSeq = dnaSeq.replace("T", "U")
    protSeq = ''
    for i in range(0,len(dnaSeq),3):
        if (i+3 > len(dnaSeq)):
            break
        protSeq += AminoAbbrv[codon[rnaSeq[i:i+3]]]
    return rnaSeq, protSeq

## Let's test it

In [5]:
# one way to code TyrocidineB1
DNA = 'GTGAAACTTTTTCCTTGGTTTAATCAATAT' 
DNAr = revComp(DNA)

print("A short DNA sequence")
print("5'-%s-3'" % DNA)
print("3'-%s-5'" % ''.join([{'A':'T','C':'G','G':'C','T':'A'}[base] for base in DNA]))
print()

print("Read frames in primary sequence")
for frame in range(3):
    RNA, Peptides = proteinTranslation(DNA[frame:]+DNA[:frame])
    print("%d  %s" % (frame+1,DNA))
    print("%s   %s" % (frame*" ", RNA))
    print("%s   %s" % (frame*" ", Peptides))
    print()
print()

print("Read frames in the reverse-complement sequence")
for frame in range(3):
    RNA, Peptides = proteinTranslation(DNAr[frame:]+DNAr[:frame])
    print("%d  %s" % (frame+1,DNAr))
    print("%s   %s" % (frame*" ", RNA))
    print("%s   %s" % (frame*" ", Peptides))
    print()

A short DNA sequence
5'-GTGAAACTTTTTCCTTGGTTTAATCAATAT-3'
3'-CACTTTGAAAAAGGAACCAAATTAGTTATA-5'

Read frames in primary sequence
1  GTGAAACTTTTTCCTTGGTTTAATCAATAT
   GUGAAACUUUUUCCUUGGUUUAAUCAAUAU
   ValLysLeuPheProTrpPheAsnGlnTyr

2  GTGAAACTTTTTCCTTGGTTTAATCAATAT
    UGAAACUUUUUCCUUGGUUUAAUCAAUAUG
    STPAsnPhePheLeuGlyLeuIleAsnMet

3  GTGAAACTTTTTCCTTGGTTTAATCAATAT
     GAAACUUUUUCCUUGGUUUAAUCAAUAUGU
     GluThrPheSerLeuValSTPSerIleCys


Read frames in the reverse-complement sequence
1  ATATTGATTAAACCAAGGAAAAAGTTTCAC
   AUAUUGAUUAAACCAAGGAAAAAGUUUCAC
   IleLeuIleLysProArgLysLysPheHis

2  ATATTGATTAAACCAAGGAAAAAGTTTCAC
    UAUUGAUUAAACCAAGGAAAAAGUUUCACA
    TyrSTPLeuAsnGlnGlyLysSerPheThr

3  ATATTGATTAAACCAAGGAAAAAGTTTCAC
     AUUGAUUAAACCAAGGAAAAAGUUUCACAU
     IleAspSTPThrLysGluLysValSerHis



## Now let's try Something Real

In [9]:
def loadFasta(filename):
    """ Parses a classically formatted and possibly 
        compressed FASTA file into a list of headers 
        and fragment sequences for each sequence contained"""
    if (filename.endswith(".gz")):
        fp = gzip.open(filename, 'r')
    else:
        fp = open(filename, 'r')
    # split at headers
    data = fp.read().split('>')
    fp.close()
    # ignore whatever appears before the 1st header
    data.pop(0)     
    headers = []
    sequences = []
    for sequence in data:
        lines = sequence.split('\n')
        headers.append(lines.pop(0))
        # add an extra "+" to make string "1-referenced"
        sequences.append('+' + ''.join(lines))
    return (headers, sequences)

header, seq = loadFasta("data/BacillusBrevis.fa")
for i in range(len(header)):
    print(header[i])
    print(len(seq[i])-1, "bases", seq[i][:30], "...", seq[i][-30:])
    print()

Chromosome dna:chromosome chromosome:GCA_000010165.1:Chromosome:1:6296436:1
6296436 bases +GGGTCTGTGGATATCATTTTATCCACAAA ... AAGGCAAATATCCCCATAAAACTATTTCCC



## Some simple "Computational" experiments

* Just like a biologist, it is ill-advised to jump into a problem without first getting some sense of what works and what doesn't. Let's first take a look at our data
* One thing you might have noticed about the possible encodings of Tyrocidine B1 is that there is a single *Tryptophan* and only one encoding for this amino acid
* Let's try to anchor our search around that seed and grow the full sequence from there.
* Where and how many *Tryptophan* encodings

In [10]:
def CodonSeqCount(genome, codonSeq):
    N = 0
    start = 0
    while True:
        pos = genome.find(codonSeq, start)
        if (pos < 0):
            break
        N += 1
        start = pos + 1
    return N
    
genome = seq[0]
tryptophanCode = TrimerCodes['W'][0].replace('U','T')
print("Searching for", tryptophanCode, "Found it", CodonSeqCount(genome,tryptophanCode), "times")
revCompCode = revComp(tryptophanCode)
print("Searching for", revCompCode, "Found it", CodonSeqCount(genome,revCompCode), "times")

Searching for TGG Found it 106555 times
Searching for CCA Found it 107112 times


Is this about what you expected? $\frac {6.3 \times 10^6}{64} \approx 100,000$.
We've narrowed our search some, but there's still alot. 

## Narrowing things down more

* Rather than searching for every possible string, let's examine the codons around this initial anchor.

In [11]:
def AminoAcidSeqCount(genome, peptideSeq):   # Note: Only checks primary strand
    readframe = []
    anchor = peptideSeq[0]
    for rnaSeq in TrimerCodes[anchor]:
        dnaSeq = rnaSeq.replace('U','T')
        start = 0
        while True:
            pos = genome.find(dnaSeq, start)
            if (pos < 0):
                break
            i = 1
            while (i < len(peptideSeq) and (pos+3*(i+1) <= len(genome))):
                if (codon[genome[pos+3*i:pos+3*(i+1)].replace('T','U')] != peptideSeq[i]):
                    break
                i += 1
            else:
                readframe.append(pos)
            start = pos + 1
        return readframe

peptide = "WFNQY"
for i in range(1,len(peptide)+1):
    ORFs = AminoAcidSeqCount(genome,peptide[:i])
    print("Searching for", peptide[:i], "Found it", len(ORFs), "times")  

Searching for W Found it 106555 times
Searching for WF Found it 3922 times
Searching for WFN Found it 111 times
Searching for WFNQ Found it 5 times
Searching for WFNQY Found it 1 times


## Could we have been Lucky this time?

Let's take a look:

In [12]:
for pos in ORFs:
    start = pos
    end = pos + (3*len(peptide))
    print("%9d %s %s" % (start, genome[start:end], proteinTranslation(genome[start:end])[1]))

  2789614 TGGTTCAACCAATAT TrpPheAsnGlnTyr


We didn't find what we were expecting.

## What Went Wrong?

1. We only searched in one direction from our anchor 
2. We didn't search the reverse-complement sequence
3. We didn't consider that the cycle could have been broken somewhere in between "W-F-N-Q-Y"
   * One of the earlier 5, 111, 3922, or 106555 candidates might be the solution
4. The best approach might be to call AminoAcidSeqCount() will all 10 circular permutations of "VKLFPWFNQY" (assuming that we also fix the reverse-complement sequence problem)

In [1]:
[TyrocidineB1[i:]+TyrocidineB1[:i] for i in range(len(TyrocidineB1))]

NameError: name 'TyrocidineB1' is not defined